fodc/proxy/internal/metrics/aggregator.go

```diff
@@ -0,0 +1,287 @@
+		collecting:  make(map[string]chan []*AggregatedMetric),
+	}
+}
+
+// SetGRPCService sets the gRPC service for sending metrics requests.
+func (ma *Aggregator) SetGRPCService(grpcService RequestSender) {
+	ma.mu.Lock()
+	defer ma.mu.Unlock()
+	ma.grpcService = grpcService
+}
+
+// ProcessMetricsFromAgent processes metrics received from an agent.
+func (ma *Aggregator) ProcessMetricsFromAgent(ctx context.Context, agentID string, agentInfo *registry.AgentInfo, req *fodcv1.StreamMetricsRequest) error {
+	aggregatedMetrics := make([]*AggregatedMetric, 0, len(req.Metrics))
+
+	for _, metric := range req.Metrics {
+		labels := make(map[string]string)
+		for key, value := range metric.Labels {
+			labels[key] = value
+		}
+
+		labels["agent_id"] = agentID
+		labels["node_role"] = agentInfo.NodeRole
+		labels["ip"] = agentInfo.PrimaryAddress.IP
+		labels["port"] = fmt.Sprintf("%d", agentInfo.PrimaryAddress.Port)
+
+		for key, value := range agentInfo.Labels {
+			labels[key] = value
+		}
+
+		var timestamp time.Time
+		switch {
+		case metric.Timestamp != nil:
+			timestamp = metric.Timestamp.AsTime()
+		case req.Timestamp != nil:
+			timestamp = req.Timestamp.AsTime()
+		default:
+			timestamp = time.Now()
+		}
+
+		aggregatedMetric := &AggregatedMetric{
+			Name:        metric.Name,
+			Labels:      labels,
+			Value:       metric.Value,
+			Timestamp:   timestamp,
+			AgentID:     agentID,
+			NodeRole:    agentInfo.NodeRole,
+			Description: metric.Description,
+		}
+
+		aggregatedMetrics = append(aggregatedMetrics, aggregatedMetric)
+	}
+
+	ma.collectingMu.RLock()
+	collectCh, exists := ma.collecting[agentID]
+	ma.collectingMu.RUnlock()
+
+	if exists {
+		select {
+		case collectCh <- aggregatedMetrics:
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			ma.logger.Warn().Str("agent_id", agentID).Msg("Metrics collection channel full, dropping metrics")
+		}
+	}
+
+	return nil
+}
+
+// CollectMetricsFromAgents requests metrics from all agents (or filtered agents) when external client queries.
+func (ma *Aggregator) CollectMetricsFromAgents(ctx context.Context, filter *Filter) ([]*AggregatedMetric, error) {
+	agents := ma.getFilteredAgents(filter)
+	if len(agents) == 0 {
+		return []*AggregatedMetric{}, nil
+	}
+
+	collectChs := make(map[string]chan []*AggregatedMetric)
+	agentIDs := make([]string, 0, len(agents))
+	ma.collectingMu.Lock()
+	for _, agentInfo := range agents {
+		collectCh := make(chan []*AggregatedMetric, 1)
+		collectChs[agentInfo.AgentID] = collectCh
+		ma.collecting[agentInfo.AgentID] = collectCh
+		agentIDs = append(agentIDs, agentInfo.AgentID)
+	}
+	ma.collectingMu.Unlock()
+
+	defer func() {
+		ma.collectingMu.Lock()
+		for _, agentID := range agentIDs {
+			delete(ma.collecting, agentID)
+		}
+		ma.collectingMu.Unlock()
+	}()
+
+	for _, agentInfo := range agents {
+		requestErr := ma.grpcService.RequestMetrics(ctx, agentInfo.AgentID, filter.StartTime, filter.EndTime)
+		if requestErr != nil {
```

Comment:
Please drop the ma.collecting entry (or close the channel) as soon as RequestMetrics returns an error so that ProcessMetricsFromAgent skips the agent immediately instead of trying to write into a dead sink.